pkg/bloomgateway/ringmanager.go

```diff
@@ -0,0 +1,264 @@
+package bloomgateway
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/kv"
+	"github.com/grafana/dskit/ring"
+	"github.com/grafana/dskit/services"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	ringAutoForgetUnhealthyPeriods = 10
+	ringNameForServer              = "bloom-gateway"
+	ringNumTokens                  = 128
+	ringCheckPeriod                = 3 * time.Second
+
+	// RingIdentifier is used as a unique name to register the Bloom Gateway ring.
+	RingIdentifier = "bloom-gateway"
+
+	// RingKey is the name of the key used to register the different Bloom Gateway instances in the key-value store.
+	RingKey = "bloom-gateway"
+)
+
+// ManagerMode defines the different modes for the RingManager to execute.
+//
+// The RingManager and its modes are only relevant if the Bloom Gateway is running in ring mode.
+type ManagerMode int
+
+const (
+	// ClientMode is the RingManager mode executed by Loki components that are clients of the Bloom Gateway.
+	// The RingManager in client will have its own ring key-value store but it won't try to register itself in the ring.
+	ClientMode ManagerMode = iota
+
+	// ServerMode is the RingManager mode execute by the Bloom Gateway.
+	// The RingManager in server mode will register itself in the ring.
+	ServerMode
+)
+
+// RingManager is a component instantiated before all the others and is responsible for the ring setup.
+//
+// All Loki components that are involved with the Bloom Gateway (including the Bloom Gateway itself) will
+// require a RingManager. However, the components that are clients of the Bloom Gateway will ran it in client
+// mode while the Bloom Gateway itself will ran the manager in server mode.
+type RingManager struct {
+	services.Service
+
+	cfg    Config
+	logger log.Logger
+
+	subservices        *services.Manager
+	subservicesWatcher *services.FailureWatcher
+
+	RingLifecycler *ring.BasicLifecycler
+	Ring           *ring.Ring
+	Mode           ManagerMode
+}
+
+// NewRingManager instantiates a new RingManager instance.
+// The other functions will assume the RingManager was instantiated through this function.
+func NewRingManager(mode ManagerMode, cfg Config, logger log.Logger, registerer prometheus.Registerer) (*RingManager, error) {
```

Comment:
nit: probably a good thing to refactor into independent code. This looks to be repeated in both the scheduler & index-gw